clickhouse/base/sslsocket.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include "socket.h"
+
+#include <memory>
+
+typedef struct ssl_ctx_st SSL_CTX;
+typedef struct ssl_st SSL;
+
+namespace clickhouse {
+
+struct SSLParams
+{
+    std::vector<std::string> path_to_ca_files;
+    std::string path_to_ca_directory;
+    bool use_default_ca_locations;
+    int context_options;
+    int min_protocol_version;
+    int max_protocol_version;
+    bool use_SNI;
+};
+
+class SSLContext
+{
+public:
+    explicit SSLContext(SSL_CTX & context);
+    explicit SSLContext(const SSLParams & context_params);
+    ~SSLContext() = default;
+
+    SSLContext(const SSLContext &) = delete;
+    SSLContext& operator=(const SSLContext &) = delete;
+    SSLContext(SSLContext &&) = delete;
+    SSLContext& operator=(SSLContext &) = delete;
+
+private:
+    friend class SSLSocket;
+    SSL_CTX * getContext();
+
+private:
+    std::unique_ptr<SSL_CTX, void (*)(SSL_CTX*)> context_;
+};
+
+class SSLSocket : public Socket {
+public:
+    explicit SSLSocket(const NetworkAddress& addr, const SSLParams & ssl_params, SSLContext& context);
+    SSLSocket(SSLSocket &&) = default;
+    ~SSLSocket() = default;
+
+    SSLSocket(const SSLSocket & ) = delete;
+    SSLSocket& operator=(const SSLSocket & ) = delete;
+
+    std::unique_ptr<InputStream> makeInputStream() const override;
+    std::unique_ptr<OutputStream> makeOutputStream() const override;
+
+private:
+    std::unique_ptr<SSL, void (*)(SSL *s)> ssl_ptr_;
+    SSL *ssl_; // for convinience with SSL API
```

Comment:
`ssl_ptr_.get()` will give you that same ptr, won't it?

Reply:
yep, fixed